src/main/java/org/spongepowered/api/event/item/inventory/UpdateAnvilEvent.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.event.item.inventory;
+
+import org.spongepowered.api.event.Cancellable;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+/**
+ * Fires whenever the left and right slots of an anvil are filled
+ * and a new result is computed.
+ *
+ * <p>Canceling this event prevents the vanilla behaviour and the
+ * output is set to empty</p>
+ *
+ * <p>Changed costs only apply when the result is also set</p>
+ */
+public interface UpdateAnvilEvent extends TargetInventoryEvent, Cancellable {
+
+    /**
+     * Returns the new item name.
+     *
+     * @return The new item name
+     */
+    String getItemName();
+
+    /**
+     * Returns the left input item.
+     *
+     * @return The left input item
+     */
+    ItemStackSnapshot getLeft();
+
+    /**
+     * Returns the right input item.
+     *
+     * @return The right input item
+     */
+    ItemStackSnapshot getRight();
+
+    /**
+     * Returns the custom result.
+     *
+     * @return The custom result or {@link Optional#empty()} when not set
+     */
+    Optional<ItemStackSnapshot> getResult();
```

Comment:
I still say this should be a transaction, because the result is always available, even if the `ItemStack` is empty, in fact, that's what the implementation will do... it doesn't deal with a null item stack in the slot.

What I'd say the transaction would look like is the following:
- `getOriginal()` would almost always be empty, or whatever is currently in the slot
- `getDefault()` would have to simulate the vanilla behavior (I know, totally separate from what Forge does, more on that later)
- `getCustom()` would be the plugin provided one, if available
- `getFinal()` would be as `Transaction` already defines, if the custom is set, returns custom, otherwise return default. 

My reasoning for this:

Forge implements this event as more of a "pre" event where the desired output can be set, and likewise the costs. My guess for the original intention was due to how anvil repairs/enchantments are not jsonified, and the event predated the start of jsonifying/configuring/registering of things like normal-esque recipes. Part of the problem here is that you're exposing directly the forge event as the forge event was written, which doesn't really provide good extensibility from a plugin's perspective (imagine if a plugin wanted to modify the costs, output, and the ingredients provided, like a reactive item that would be "locked" or modified prior to the transaction. I know, it's a little awkward since the event is intended to be thrown whenever the anvil *container* is being told to refresh, which can be a little spammy.

I personally think we can and should be able to represent the playback of the information, whether it's done through an overwrite, or whether we do a couple of injectors/redirects.